temporalio/contrib/opentelemetry.py

```diff
@@ -0,0 +1,439 @@
+            return await super().signal_workflow(input)
+
+
+class _TracingActivityInboundInterceptor(temporalio.worker.ActivityInboundInterceptor):
+    def __init__(
+        self,
+        next: temporalio.worker.ActivityInboundInterceptor,
+        root: TracingInterceptor,
+    ) -> None:
+        super().__init__(next)
+        self.root = root
+
+    async def execute_activity(
+        self, input: temporalio.worker.ExecuteActivityInput
+    ) -> Any:
+        info = temporalio.activity.info()
+        with self.root.tracer.start_as_current_span(
+            f"RunActivity:{info.activity_type}",
+            context=self.root._context_from_headers(input.headers),
+            attributes={
+                "temporalWorkflowID": info.workflow_id,
+                "temporalRunID": info.workflow_run_id,
+                "temporalActivityID": info.activity_id,
+            },
+        ):
+            return await super().execute_activity(input)
+
+
+class _WorkflowExternFunctions(TypedDict):
+    __temporal_opentelemetry_new_completed_span: Callable[
+        [_CarrierDict, str, opentelemetry.util.types.Attributes, int], _CarrierDict
+    ]
+
+
+class TracingWorkflowInboundInterceptor(temporalio.worker.WorkflowInboundInterceptor):
+    """Tracing interceptor for workflow calls.
+
+    See :py:class:`TracingInterceptor` docs on why one might want to subclass
+    this class.
+    """
+
+    def __init__(self, next: temporalio.worker.WorkflowInboundInterceptor) -> None:
+        """Initialize a tracing workflow interceptor."""
+        super().__init__(next)
+        self._extern_functions = cast(
+            _WorkflowExternFunctions, temporalio.workflow.extern_functions()
+        )
+        # To customize these, like the primary tracing interceptor, subclassing
+        # must be used
+        self.header_key: str = "_tracer-data"
+        self.text_map_propagator: opentelemetry.propagators.textmap.TextMapPropagator = (
+            default_text_map_propagator
+        )
+        # TODO(cretz): Should I be using the configured one for this workflow?
+        self.payload_converter = temporalio.converter.default().payload_converter
+
+    def init(self, outbound: temporalio.worker.WorkflowOutboundInterceptor) -> None:
+        """Implementation of
+        :py:meth:`temporalio.worker.WorkflowInboundInterceptor.init`.
+        """
+        super().init(_TracingWorkflowOutboundInterceptor(outbound, self))
+
+    async def execute_workflow(
+        self, input: temporalio.worker.ExecuteWorkflowInput
+    ) -> Any:
+        """Implementation of
+        :py:meth:`temporalio.worker.WorkflowInboundInterceptor.execute_workflow`.
+        """
+        with self._instrument(
+            f"RunWorkflow:{temporalio.workflow.info().workflow_type}",
+            parent_headers=input.headers,
+        ):
+            return await super().execute_workflow(input)
+
+    async def handle_signal(self, input: temporalio.worker.HandleSignalInput) -> None:
```

Comment:
We need to have discussions about what the hierarchy is for signals here. So right now, the parent of everything that happens in the signal is the span created on client outbound. That is a top-level span, it is _not_ tied to the workflow in any way.

Is that what we want? A "client signal" -> "workflow signal" -> "all the stuff that happened in the handler" trace? Or do we want things that happen in signal handlers to be children of the workflow span?

Reply:
IMO some of these should be "follows from" relationships, not children. The stuff that happens within a workflow should be a child of that workflow, but I think the workflow itself should "follow from" the client execute, and things that happen inside a signal handler should probably "follow from" that signal span (though often their start time will be within the signal span).

Reply:
:+1: Gonna put this back in draft and rework this

Reply:
Reworked.